esp/smc/SMCLib/InfoCacheReader.hpp

```diff
@@ -0,0 +1,128 @@
+const unsigned DEFAULTINFOCACHEAUTOREBUILDSECOND = 120;
+
+class CInfoCache : public CSimpleInterfaceOf<IInterface>
+{
+protected:
+    CDateTime timeCached;
+public:
+
+    CInfoCache() {};
+
+    virtual bool isCachedInfoValid(unsigned timeOutSeconds);
+    virtual inline StringBuffer& queryTimeCached(StringBuffer& str) { return timeCached.getString(str, true); }
+};
+
+interface IInfoCacheReader : extends IInterface
+{
+    virtual CInfoCache* read() = 0;
+};
+
+class CInfoCacheReaderThread : public CSimpleInterfaceOf<IThreaded>
+{
+    StringAttr name;
+    bool stopping = false;
+    bool detached = false;
+    bool first = true;
+    bool firstBlocked = false;
+    unsigned autoRebuildSeconds = DEFAULTINFOCACHEAUTOREBUILDSECOND;
+    unsigned forceRebuildSeconds = DEFAULTINFOCACHEFORCEBUILDSECOND;
+    Owned<CInfoCache> infoCache;
+    Owned<IInfoCacheReader> infoCacheReader;
+    Semaphore sem;
+    Semaphore firstSem;
+    CriticalSection crit;
+    CThreaded threaded;
+    std::atomic<bool> waiting = {false};
+
+public:
+    CInfoCacheReaderThread(IInfoCacheReader *_infoCacheReader, const char* _name, unsigned _autoRebuildSeconds, unsigned _forceRebuildSeconds)
+        : name(_name), autoRebuildSeconds(_autoRebuildSeconds), forceRebuildSeconds(_forceRebuildSeconds), threaded(_name)
+    {
+        infoCacheReader.setown(_infoCacheReader);
+        threaded.init(this);
+    };
+
+    ~CInfoCacheReaderThread()
+    {
+        stopping = true;
+        sem.signal();
+        firstSem.signal(); // in case
+        threaded.join();
+    }
+
+    virtual void threadmain() override;
+    CInfoCache* getCachedInfo()
+    {
+        CLeavableCriticalBlock b(crit);
+        if (first)
+        {
+            if (detached)
+                return nullptr;
+            firstBlocked = true;
+            b.leave();
+            firstSem.wait();
+            b.enter();
+            if (first)
+                return nullptr;
+        }
+
+        //Now, activityInfoCache should always be available.
+        assertex(infoCache);
+        if (!detached && !infoCache->isCachedInfoValid(forceRebuildSeconds))
+            rebuild();
+        return infoCache.getLink();
+    }
+    void rebuild()
+    {
+        bool expected = true;
+        if (waiting.compare_exchange_strong(expected, false))
+            sem.signal();
+    }
+    void setDetachedState(bool _detached)
+    {
+        CriticalBlock b(crit);
+        if (detached != _detached)
+        {
+            detached = _detached;
+            if (detached)
+            {
+                // NB: first will still be true, signal and getActivityInfo() will return null
+                if (firstBlocked)
+                {
+                    firstBlocked = false;
+                    firstSem.signal(); // NB: first still true
+                }
+            }
+            else
+                rebuild();
+        }
+    }
+    bool isDaliDetached() const { return detached; }
```

Comment:
minor: I think isActive() would be a bit clearer in this context.

Reply:
Will change here and other similar place.